Improve filtering and performance for icon picker

steverep opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Follow up to #14387 per my comments to fix by moving the filter computation code so it is only done once per update.  Note the comment I made regarding loading custom icon sets which might also cause problems.  Feel free to push tweaks to this branch.

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #14147, fixes #14462
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

piitaya committed:
Avoid browser freeze when copy paste icon in icon picker

piitaya committed:
Clean ha-combobox events

piitaya committed:
use state instead of combobox assignment

steverep committed:
Move icon filtering to render

bramkragten commented on src/components/ha-icon-picker.ts:
```diff
@@ -48,17 +47,41 @@ export class HaIconPicker extends LitElement {
+  @state() private _filterString = "";
 
   protected render(): TemplateResult {
+    const characterCount = this._filterString.length;
+    let filteredItems: IconItem[] = [];
+    if (characterCount >= 2) {
+      const filteredItemsByKeywords: IconItem[] = [];
+
+      iconItems.forEach((item) => {
+        if (item.icon.includes(this._filterString)) {
+          filteredItems.push(item);
+          return;
+        }
+        if (item.keywords.some((t) => t.includes(this._filterString))) {
+          filteredItemsByKeywords.push(item);
+        }
+      });
+
+      filteredItems.push(...filteredItemsByKeywords);
+
+      if (!filteredItems.length) {
+        filteredItems = [{ icon: this._filterString, keywords: [] }];
+      }
+    } else {
+      filteredItems = iconItems;
+    }
```
Can we move this to a memoized function that takes `_filterString` and `iconItems`?

steverep replied:
I suppose.  Would that risk over-caching if there were a lot of different filter strings attempted?

steverep replied:
Done.  Forgot it's only caching the last.

steverep committed:
Await all custom icon loads before adding

steverep committed:
Memoize filter and remove character count

steverep commented:
Performance on the filter seems fine to me both on Firefox and Chrome.  Not sure if the web worker would gain much but extra code.

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -87,46 +87,77 @@ export class HaIconPicker extends LitElement {
-        this._loadCustomIconItems(iconSet);
+        customIconLoads.push(this._loadCustomIconItems(iconSet));
+      });
+      (await Promise.all(customIconLoads)).forEach((customIconItems) => {
```
If a promise doesn't resolve, no icons will be added.

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -58,7 +58,7 @@ export class HaIconPicker extends LitElement {
         .value=${this._value}
         allow-custom-value
-        .filteredItems=${iconItems}
+        .filteredItems=${this._filterIcons(this._filterString)}
```
```suggestion
        .filteredItems=${this._filterIcons(this._filterString, ICONS)}
```
Otherwise, the items are not updated when the icons are loaded.

steverep replied:
Harmless change, but I don't understand why that makes a difference?

piitaya replied:
`_filterIcons` is a memoized function. Once icons are loaded, `requestUpdate` is call but the parameters of `_filterIcons` are the same because `ICONS` are not passed as parameter.

steverep replied:
So the default for the parameter won't be checked by `memoizeOne`?

steverep replied:
Hmm... just tested.  I guess not.  Good to know.

piitaya replied:
No it only checks parameters passed to the function.

piitaya commented:
The changes doesn't fix the issue 😕
To reproduce : 
- enter "mdi:home"
- copy the text
- delete the text
- paste the text 
- freeze 😕

steverep commented:
I just tried those steps and I can't get it to freeze.  I've tried pasting strings of varying length, typing gibberish fast, etc.  Are you sure the browser cache was completely cleared?  If so, what versions of stuff?  I'm on Windows 10 with Chrome and/or Firefox (all up to date).

steverep committed:
Add icons parameter to filter call

Co-authored-by: Paul Bottein <paul.bottein@gmail.com>

piitaya commented:
I have the issue with Safari and Chrome (last version) on Mac OS (last version)

steverep commented:
Ugh... Did you happen to test if the `vaadin-combo-box` demos or other HA pickers have the same issue?

steverep commented:
Assuming we're close... can I get rid of `shouldUpdate`?  Right now it prevents initial filtering, i.e.:
1. Enter a character into an empty field before icons are loaded.
2.  You get no results because the first update goes forward with the new filter string but no icons, then another should pass with the `requestUpdate` call but doesn't because `shouldUpdate` says no when it's open.

piitaya commented:
Yes, I tested it and I only have the issue with the icon picker. I don't have the issue using web worker.

steverep commented:
Yeah was just trying to fix the initial filter not working.  Won't affect the freeze.  I don't understand that other issue.  Why would `hass` affect the filter?

In any case, I removed it and tested on iOS with no issue.  Please confirm.

I was having issues with the web worker version.  Let me go back and see if I can get repeatable steps.

steverep committed:
Fix issue with no filter results on first open

piitaya commented:
> Yeah was just trying to fix the initial filter not working. Won't affect the freeze. I don't understand that other issue. Why would `hass` affect the filter?

Before we used `combobox.filteredItems` to update the result. When `hass` was updated, it triggered rendering and it reset the results. But now we are using reactive state so this problem is solved.

steverep committed:
Guard the filter computation

steverep commented:
Just in case other properties are triggering extra filtering, I used the `guard` directive on it.  See if that solves it on your Mac.

Otherwise I guess we should combine these techniques with the web worker or optimize another way.  When it was crashing I couldn't even use developer tools to profile it without freezing that up too. 😞

steverep committed:
Add filter stats to log

steverep committed:
Fix duplicate filter results and use Array.filter

steverep commented:
Okay I added some code to log the filter stats:

- Noticed that some searches produced more results than total icons, because the keyword search is done as a repeat and then just pushed to the array.  So I fixed that just by using `.filter()` once, which is also significantly faster.
- There are 7096 icons and all my searches take 1-3 ms (about 1/2 of the time using `foreach`/`push`), with single characters really only at the 3 ms mark.

Given those puny timings, I'm thinking the web worker "working" was more just luck?  And it didn't help on my machine at all.  This PR has definitely improved performance (and fixed the crashes for me) but something else is funky.

steverep committed:
Use performance with a decimal instead of Date

steverep committed:
Schedule updates only once per frame

steverep commented:
@piitaya try this now.  It's the same change you made to await the next paint, but just does it by slowing down the whole update cycle.

steverep commented:
😰 🤨

steverep committed:
Remove unnecessary opened state and prevent missing custom icons

steverep committed:
Move icon loading code outside element

steverep committed:
Merge dev to get vaadin bump

steverep commented:
@piitaya if it's okay with you, consider merging this branch anyway even though it doesn't fix the freeze per se.  It includes a number of other fixes/performance enhancements:

- Combine icon name and keywords searches into one filter to prevent duplicate results and cut time in half
- Only perform filter once per update (instead of once per `filter-changed` event, which could fire more than once before the update starts)
- Only repeat filter search when filter string changes (via guard directive)
- Await custom icon set loads to prevent racing pushes (also ensures they are listed in order)
- Ensure custom icons are in initial filters
- Housekeeping to use reactive state, remove character count, and clean up combo-box events (your commits)

The code to slow the rendering down to once per frame is probably not needed (unless you are noticing some jank while typing a filter string).  Although if that's the case it should probably be included in the combo-box element instead.

I really think this is something weird in the vaadin code, probably in the stuff they ported from polymer.  I think we're just getting lucky/unlucky on fixing it.  I can file an issue with them.

Let me know what you think.

steverep committed:
Merge dev to get TypeScript upgrade

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -87,46 +131,37 @@ export class HaIconPicker extends LitElement {
+          item.icon.includes(filterString) ||
+          item.keywords.some((word) => word.includes(filterString))
+      );
+      if (__DEV__) {
```
It's better to remove this. We do not want to overloaded the console in dev.

steverep replied:
Yeah I can get rid of it, but it might be useful if we start messing with a better rank and sort.

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -87,46 +131,37 @@ export class HaIconPicker extends LitElement {
+        return iconItems;
+      }
+      const startTime = performance.now();
+      const filteredItems = iconItems.filter(
+        (item) =>
+          item.icon.includes(filterString) ||
+          item.keywords.some((word) => word.includes(filterString))
+      );
```
We didn't use filter because we wanted to display the matches by name before the matches by keyword.
If you type `nature` you prefer to have `mdi:nature` instead of `mdi:bee`

steverep replied:
My primary goal was to get rid of the duplicates and have one loop instead of two for performance.  For example, when you type "m" it searches 7096 icons and finds over 10k results.

As for ranking the includes over the keywords, I think that's flawed logic, especially when searching for words that are commonly part of other words.  For example, if I search for "rain", the icons I probably care about are buried below a whole bunch of "train" icons.  Or if I search for "hot", there are a bunch of screenshot icons to contend with before getting to the relevant ones.

I think trying to rank and sort them is a good idea, but the algorithm needs to be more nuanced and it should use a sorting function after the filter for good performance.

steverep replied:
I implemented a simple ranking when filtering.  This restores your preference to rank icon name matches over keywords, but also addresses the issues I pointed out by ranking exact matches higher than partial word inclusions.  Test it out and let me know what you think.

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -46,9 +85,12 @@ export class HaIconPicker extends LitElement {
+  @state() private _filterString = "";
 
-  @query("ha-combo-box", true) private comboBox!: HaComboBox;
+  protected override async scheduleUpdate() {
+    await nextRender();
+    super.scheduleUpdate();
+  }
```
We should add comment to explain why we do this.

steverep replied:
Will do.

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -58,7 +100,9 @@ export class HaIconPicker extends LitElement {
         .value=${this._value}
         allow-custom-value
-        .filteredItems=${iconItems}
+        .filteredItems=${guard([this._filterString, ICONS.length], () =>
+          this._filterIcons(this._filterString, ICONS)
+        )}
```
Are you sure the guard and the memoize function is needed ? Only one of these should be necessary I think.

steverep replied:
Yeah they are mostly redundant.  The `guard` has the extra advantage of protecting the template itself.  This will be a moot point when I submit the actual fix. 😉

piitaya commented:
Your changes fixes the issue for me thank to the `await nextRender()` line. Adding this line in the `filterChanged` function worked for me too but I'm ok if should prefer to do the result calculation in the render.

steverep commented:
So after those replies, here's the good and bad news...

Bad news is that this branch no longer works for me to prevent crashes. The upgrade of vaadin somehow broke it. 😭 

Good news is that I think I found an _actual_ fix which I have locally branched off this PR but won't have time to push for a couple days because of the holiday here. 🦃

steverep committed:
Use data provider callback from vaadin API

steverep commented:
Sorry for being cryptic - it's been a hectic few days for me.

I pushed a change that now uses the `dataProvider` callback instead of handing the `filteredItems` to the combo-box, which fixes the issue again for me regardless of vaadin version.  Please tell me it still works for you too. 🤞 

I deduced that the only thing that could be causing the issue for the icon picker and not the others was the number of items, and also found vaadin/web-components#874, which certainly sounds like the same issue.  Looking at their code, although the overlay may actually only display ~10 items at a time, it calls the `renderer` on all the items provided, which would consume a big chunk of resources.  So pasting an icon name would start by rendering 7096 icons on open then immediately clear 7095 of them.  Hence the unpredictable results.

Using the `dataProvider` paginates the items so the same operation would only be clearing `pageSize` items (defaults to 50).  It also makes listening for filter changes unnecessary since the data provider handles it.

steverep committed:
Add comment to explain await next render

steverep committed:
Remove stats from dev log

piitaya reviewed (commented):
I didn't have the issue with your changes ! 🎉 Good job ! 💪
The `inert` attributes removal doesn't work anymore as we are not using `items` and `filteredItems`. I pushed a fix for this : https://github.com/home-assistant/frontend/pull/14463.

I see two changes compared to the old icon picker :
1. With the old filter, we returned `[{ icon: filterString, keywords: [] }]` there was no result to show the user a preview of their custom icon (in the case the custom icon is not listed).
2. Can we re-test with the old `filterIcon` function now the bug is fixed to keep the priority feature (name matches before keywords matches)?

steverep committed:
Merge dev to get inert removal fix

steverep committed:
Implement simple rank and sort algorithm

piitaya commented:
We already use `Fuse.js` for some searches. May be we can re-use it here ?
https://fusejs.io/

steverep committed:
Allow preview for custom icon not in lists

steverep committed:
Remove unnecessary await for next render

steverep commented:
I think I addressed all of your comments, with a tweak to the ranking per my comment a few minutes ago.  I can take a look at that package but probably not until a bit later tonight.

I also removed the `scheduleUpdate` override because this element is not controlling the filter calls anymore (combo-box controls it through `dataProvider`).

steverep commented:
Fuse looks like it could work, but it's not immediately clear to me the best way to implement it.  It would take some extra testing and we should time it for comparison.  Maybe save that for another PR?

piitaya commented on src/components/ha-icon-picker.ts:
```diff
@@ -87,48 +130,57 @@ export class HaIconPicker extends LitElement {
-    if (this._opened && !iconLoaded) {
-      const iconList = await import("../../build/mdi/iconList.json");
+  // Filter can take a significant chunk of frame (up to 3-5 ms)
+  private _filterIcons = memoizeOne(
+    (filter: string, iconItems: IconItem[] = ICONS) => {
```
It's a filter so it should return `IconItem[]` for `RankedIcon[]`

steverep replied:
Not sure what change you are suggesting.  It returns `RankedIcon[]` unless the filter string is empty, in which case it returns `IconItem[]` just to avoid the expense of creating a new array.  And in the filtered case, it just avoids the unnecessary copies of `parts` and `keywords`.

steverep replied:
@piitaya do you want to get some beta testing for this?  I can make final tweaks but I don't know what you are requesting from your comment.

piitaya reviewed (commented):
It looks good to me. 
I only left a comment about the filter 🙂
And I agree, we can test `Fuse` in another PR. (And I'm not sure we will improve performance).

bramkragten commented on src/components/ha-icon-picker.ts:
```diff
@@ -87,48 +130,57 @@ export class HaIconPicker extends LitElement {
+      // Filter and rank such that exact matches rank higher, and prefer icon name matches over keywords
+      for (const item of iconItems) {
+        if (item.parts.has(filter)) {
+          addIcon(item.icon, 1);
```
I think adding the size of `parts` to the ranking might make it a bit better.

The when you search `tree`, `mdi:tree` will be on top and not `mdi:chart-tree` (or whatever)

steverep replied:
Yeah that's a good idea.  You'd have to loop manually over `parts` and set the rank to something like `indexFound/item.parts.size`.

I'm afraid to slow it down any further though because the freezing is clearly affected by the timing.  If I play around with the picker enough in a session, I can actually still get Chrome to freeze, but it's not consistent and much much more difficult.  There's definitely an issue in the vaadin code.
